In [16]:
# This code is developed by Pooyan :D

# What is our goal?
In this file we will try to decrypt the cipher text by Babbage-Kasiski test.

# What is Babbage-Kasiski test?
The Kasiski examination involves __looking for strings of characters that are repeated in the ciphertext__. The strings should be three characters long or more for the examination to be successful. Then, the distances between consecutive occurrences of the strings are likely to be multiples of the length of the keyword.

# How we work?
First, we will calculate the number of three letters combination. After that we will start Babbage-Kasiski test.

In [17]:
import math

f = open('cipherText.txt', 'r')
txt = f.read()
print(txt.__len__())

1378


In [18]:
threeLetters = {}
for i in range(2, 1378):
    st = txt[i-2] + txt[i-1] + txt[i]
    if threeLetters.get(st) is not None:
        threeLetters.__setitem__(st, threeLetters.get(st) + 1)
    else:
        threeLetters.__setitem__(st, 1)
print(threeLetters)

{'DBS': 1, 'BSR': 1, 'SRO': 1, 'ROS': 3, 'OSI': 3, 'SIH': 2, 'IHV': 6, 'HVC': 6, 'VCT': 4, 'CTZ': 1, 'TZD': 1, 'ZDG': 1, 'DGH': 1, 'GHR': 3, 'HRI': 4, 'RIZ': 1, 'IZE': 1, 'ZEC': 1, 'ECF': 1, 'CFR': 1, 'FRA': 1, 'RAA': 5, 'AAI': 2, 'AIH': 1, 'IHB': 1, 'HBC': 2, 'BCW': 1, 'CWG': 1, 'WGU': 1, 'GUS': 1, 'USO': 1, 'SOR': 1, 'ORU': 1, 'RUE': 1, 'UET': 1, 'ETG': 1, 'TGH': 2, 'GHD': 1, 'HDO': 3, 'DOE': 3, 'OEI': 3, 'EIH': 2, 'HVG': 2, 'VGS': 2, 'GSG': 3, 'SGT': 1, 'GTR': 1, 'TRW': 1, 'RWR': 1, 'WRI': 1, 'RIB': 2, 'IBC': 3, 'BCH': 3, 'CHW': 1, 'HWQ': 1, 'WQT': 2, 'QTG': 2, 'TGW': 3, 'GWS': 5, 'WSH': 2, 'SHS': 1, 'HSS': 2, 'SSR': 1, 'SRI': 1, 'RIC': 1, 'ICT': 1, 'CTR': 2, 'TRS': 1, 'RSN': 3, 'SNV': 3, 'NVS': 4, 'VSH': 3, 'SHD': 1, 'EIQ': 1, 'IQF': 3, 'QFW': 2, 'FWP': 2, 'WPG': 2, 'PGD': 2, 'GDU': 2, 'DUF': 2, 'UFY': 2, 'FYP': 2, 'YPU': 2, 'PUX': 2, 'UXQ': 2, 'XQL': 1, 'QLC': 1, 'LCX': 1, 'CXN': 1, 'XNB': 1, 'NBD': 1, 'BDZ': 3, 'DZC': 2, 'ZCS': 1, 'CSG': 2, 'SGP': 1, 'GPB': 1, 'PBR': 2, 'BRR': 1,

In [19]:
sorted(threeLetters.items(), key=lambda x : x[1], reverse=True)

[('IHV', 6),
 ('HVC', 6),
 ('RAA', 5),
 ('GWS', 5),
 ('ASH', 5),
 ('VCT', 4),
 ('HRI', 4),
 ('NVS', 4),
 ('GRA', 4),
 ('GIV', 4),
 ('HGR', 4),
 ('RAG', 4),
 ('IOB', 4),
 ('HWL', 4),
 ('TPP', 4),
 ('GHS', 4),
 ('SDR', 4),
 ('HRH', 4),
 ('RHR', 4),
 ('ROS', 3),
 ('OSI', 3),
 ('GHR', 3),
 ('HDO', 3),
 ('DOE', 3),
 ('OEI', 3),
 ('GSG', 3),
 ('IBC', 3),
 ('BCH', 3),
 ('TGW', 3),
 ('RSN', 3),
 ('SNV', 3),
 ('VSH', 3),
 ('IQF', 3),
 ('BDZ', 3),
 ('ITC', 3),
 ('ETH', 3),
 ('RCH', 3),
 ('OBB', 3),
 ('BBT', 3),
 ('HAY', 3),
 ('RCA', 3),
 ('CAN', 3),
 ('ANU', 3),
 ('NUG', 3),
 ('UGT', 3),
 ('GTF', 3),
 ('WLG', 3),
 ('LGG', 3),
 ('HMG', 3),
 ('TZP', 3),
 ('ZPH', 3),
 ('RSG', 3),
 ('HOU', 3),
 ('OUC', 3),
 ('UCT', 3),
 ('PPB', 3),
 ('PBH', 3),
 ('AOA', 3),
 ('SHM', 3),
 ('FGR', 3),
 ('AAS', 3),
 ('TRA', 3),
 ('BLB', 3),
 ('THH', 3),
 ('HHF', 3),
 ('WBE', 3),
 ('FIO', 3),
 ('SGX', 3),
 ('GXB', 3),
 ('XBH', 3),
 ('TGD', 3),
 ('HVR', 3),
 ('SIH', 2),
 ('AAI', 2),
 ('HBC', 2),
 ('TGH', 2),
 ('EIH', 2),

In [20]:
indOfIHV = [0] * 6
start = 0
for i in range(6):
    indOfIHV[i] = txt.index('IHV', start)
    print("start = ", start)
    start = indOfIHV[i] + 3
    print(indOfIHV[i])

start =  0
6
start =  9
42
start =  45
516
start =  519
906
start =  909
1158
start =  1161
1170


In [21]:
indOfIHV

[6, 42, 516, 906, 1158, 1170]

The ciphertext string IHV occurs in six places in the ciphertext, beginning at positions 6, 42, 516, 906, 1158, and 1170. The distances from the first occurrence to the other four occurrences are (respectively) 36, 510, 900, 1152, and 1164.

In [22]:
size = math.gcd(36, 510, 900, 1152, 1164)
size

6